//
//   Copyright 2013 The Trustees of Princeton University
//   All Rights Reserved
//

package Serialization;

message FileMetadata {
   required string fs_path = 9;
   required uint64 volume_id = 1;
   required uint64 coordinator_id = 2;
   required uint64 file_id = 3;
   required int64 file_version = 4;
   required int64 size = 5;
   required int64 mtime_sec = 6;
   required int32 mtime_nsec = 7;
   required int64 write_nonce = 8;
}

message BlockList {
   required uint64 start_id = 1;
   required uint64 end_id = 2;
   repeated int64 version = 3;
}

// truncate a file
message TruncateRequest {
   required uint64 volume_id = 1;
   required uint64 coordinator_id = 2;
   required uint64 file_id = 3;
   required int64 file_version = 4;
   required int64 size = 5;
   required string fs_path = 6;
}

// detach a file or directory
message DetachRequest {
   required uint64 volume_id = 1;
   required uint64 coordinator_id = 2;
   required uint64 file_id = 3;
   required int64 file_version = 4;
   required string fs_path = 5;
}

// block url set (for manifests)
message BlockURLSetMsg {
   required uint64 gateway_id = 1;              // if this is a cert bundle, then this is the gateway ID
   required uint64 start_id = 2;                // if this is a cert bundle, then this is the gateway type
   required uint64 end_id = 3;                  // if this is a cert bundle, then this is the gateway's caps
   repeated int64 block_versions = 4;
}

// manifest
message ManifestMsg {
   required uint64 volume_id = 1;
   required uint64 coordinator_id = 2;      // if 0, this is a cert bundle from the MS
   required uint64 file_id = 3;
   required int64 file_version = 4;
   required int64 size = 5;             // if gateway_id == 0, then this is simply the number of certificates
   required int64 mtime_sec = 6;
   required int32 mtime_nsec = 7;
   repeated BlockURLSetMsg block_url_set = 8;
   required string signature = 9;       // signature of this message using the origin gateway's private key
}

// accepted message
message AcceptMsg {
   required string fs_path = 7;
   required uint64 volume_id = 1;
   required uint64 coordinator_id = 2;
   required uint64 file_id = 3;
   required int64 file_version = 4;
   repeated uint64 block_id = 5;
   repeated int64 block_version = 6;
}

// PREPARE message uses FileMetadata and BlockList
// PROMISE message uses nothing--it's an acknowledgement to a PREPARE message
// ACCEPTED message uses AcceptMsg
// TRUNCATE message uses TruncateRequest and BlockList
// DETACH message uses DetachRequest

message WriteMsg {
   
   enum MsgType {
      PREPARE = 1;
      PROMISE = 2;
      ACCEPT = 3;
      ACCEPTED = 4;
      TRUNCATE = 5;
      DETACH = 6;

      ERROR = 7;
   }

   required uint64 volume_version = 1;    // what is the known volume metadata version on the sender
   required uint64 cert_version = 2;      // what is the known certificate bundle version on the sender
   
   required MsgType type = 3;             // what type of message
   required uint64 user_id = 4;           // what user is running this gateway
   required uint64 volume_id = 5;         // what volume is the sender in
   required uint64 gateway_id = 6;        // who's sending this message

   optional FileMetadata metadata = 7;
   optional BlockList blocks = 8;
   optional TruncateRequest truncate = 9;
   optional DetachRequest detach = 10;
   optional AcceptMsg accepted = 11;

   optional int32 errorcode = 12;
   optional string errortxt = 13;

   required string signature = 14;     // cryptographic signature with the gateway's private key
}

